cpp/src/copying/gather_opt.cu

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gather.hpp"
+#include <cudf/copying.hpp>
+#include <cudf/cudf.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/cudf_utils.h>
+#include <utilities/type_dispatcher.hpp>
+#include <bitmask/legacy_bitmask.hpp>
+#include <cudf/table.hpp>
+#include <string/nvcategory_util.hpp>
+
+#include <algorithm>
+#include <thrust/gather.h>
+
+#include <table/device_table.cuh>
+
+#include <cudf/table.hpp>
+
+#include <utilities/column_utils.hpp>
+
+/**
+ * @brief Operations for copying from one column to another
+ * @file copying_ops.cu
+ */
+
+namespace impl {
+
+template<class T>
+__device__ inline void warp_wise_reduce(T& f) {
+  #pragma unroll
+  for(int offset = 16; offset > 0; offset /= 2){
+    // TODO: ONLY works for CUDA 9.2 or later
+    float other_f = __shfl_down_sync(0xffffffffu, f, offset);
+    f += other_f;
+  }
+}
+
+template<class T>
+__device__ inline void block_wise_reduce(T f, T* smem){
```

Comment:
I think you should just use CUB for block reductions. Reduces code maintenance and bugs, and its generally very fast. Also makes it more consistent across modules -- we should try not to duplicate common code like reductions.